File: ios/NativeOperationsModule.swift
//
//  NativeOperationsModule.swift
//  NativeOperations
//
//  Created by Theo Madzou on 29/01/2025.
//

import Foundation
import React
import Swoir
import SwoirCore
import Swoirenberg
import BigInt

@objc(NativeOperationsModule)
class NativeOperationsModule: NSObject {
  // Cache for recently computed trees (LRU with max 10 trees)
  private var treeCache: [String: (tree: IMT, leaves: [String], height: Int)] = [:]
  private let maxCacheSize = 10
  private let cacheQueue = DispatchQueue(label: "merkle.cache", attributes: .concurrent)
  @objc(computeMerkleProof:index:height:resolve:reject:)
  func computeMerkleProof(_ leaves: [String], index: Int, height: Int, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
    // Run computation on background queue to avoid blocking UI
    DispatchQueue.global(qos: .userInitiated).async {
      do {
        // Create cache key
        let cacheKey = "\(height):\(leaves.joined(separator: ","))"
        
        let tree: IMT = self.cacheQueue.sync {
          if let cached = self.treeCache[cacheKey],
             cached.leaves == leaves && cached.height == height {
            // Move to end (LRU) by removing and re-adding
            self.treeCache.removeValue(forKey: cacheKey)
            self.treeCache[cacheKey] = cached
            return cached.tree
          }
          
          // Not in cache, create new tree
          let newTree = IMT(hashFunction: Poseidon2.hash, height: height)
          
          // Add to cache (remove oldest if necessary)
          if self.treeCache.count >= self.maxCacheSize {
            let oldestKey = self.treeCache.keys.first!
            self.treeCache.removeValue(forKey: oldestKey)
          }
          self.treeCache[cacheKey] = (tree: newTree, leaves: leaves, height: height)
          return newTree
        }
        
        // Convert hex strings to BigInts
        let leafBigInts = leaves.map { BigInt($0.replacingOccurrences(of: "0x", with: ""), radix: 16)! }
        // Initialize tree with leaves
        tree.initialize(zeroValue: BigInt(0), leaves: leafBigInts)
        // Create proof for the given index
        let proof = try tree.createProof(index: index)
        // Convert siblings to hex strings
        let path = proof.siblings.map { "0x" + String($0, radix: 16).padding(toLength: 64, withPad: "0", startingAt: 0) }
        let root = "0x" + String(proof.root, radix: 16).padding(toLength: 64, withPad: "0", startingAt: 0)
        
        // Return result on main queue
        DispatchQueue.main.async {
          resolve(["path": path, "index": index, "root": root])
        }
      } catch {
        print("Error", error)
        DispatchQueue.main.async {
          reject("MERKLE_PROOF_ERROR", "Error computing the merkle proof", error)
        }
      }
    }
  }
}

File: android/app/src/main/java/app/zkpassport/zkpassport/NativeOperationsModule.kt
package app.zkpassport.zkpassport

import com.facebook.react.bridge.NativeModule
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableArray
import java.util.Map
import java.util.HashMap
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.WritableMap
import com.facebook.react.bridge.WritableArray
import com.facebook.react.bridge.Arguments
import androidx.annotation.NonNull
import android.net.Uri
import com.facebook.react.bridge.Promise
import java.io.IOException
import android.content.Context
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.FileOutputStream
import android.util.Log
import java.math.BigInteger
import kotlinx.coroutines.*
import java.util.concurrent.Executors

class NativeOperationsModule(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {
    override fun getName() = "NativeOperationsModule"
    
    // Use a shared thread pool for better resource management
    private val computationScope = CoroutineScope(
        Dispatchers.Default + SupervisorJob()
    )
    
    // Cache for recently computed trees (LRU with max 10 trees)
    private val treeCache = linkedMapOf<String, Triple<MerkleTree, List<String?>, Int>>()
    private val maxCacheSize = 10

    @ReactMethod fun computeMerkleProof(leaves: ReadableArray, index: Int, height: Int, promise: Promise) {
        computationScope.launch {
            try {
                // Convert leaves array to List<String?> efficiently
                val leafList = (0 until leaves.size()).map { leaves.getString(it) }
                
                // Create cache key
                val cacheKey = "$height:${leafList.joinToString(",")}"
                
                val merkleTree = synchronized(treeCache) {
                    // Check if we have a valid cached tree
                    treeCache[cacheKey]?.let { (cachedTree, cachedLeaves, cachedHeight) ->
                        if (cachedLeaves == leafList && cachedHeight == height) {
                            // Move to end (LRU)
                            treeCache.remove(cacheKey)
                            treeCache[cacheKey] = Triple(cachedTree, cachedLeaves, cachedHeight)
                            return@synchronized cachedTree
                        }
                    }
                    
                    // Not in cache, compute new tree
                    val newTree = MerkleTree(height)
                    
                    // Add to cache (remove oldest if necessary)
                    if (treeCache.size >= maxCacheSize) {
                        val oldestKey = treeCache.keys.first()
                        treeCache.remove(oldestKey)
                    }
                    treeCache[cacheKey] = Triple(newTree, leafList, height)
                    newTree
                }
                
                // Filter out null values and compute proof asynchronously without blocking
                val nonNullLeafList = leafList.filterNotNull()
                val proof = merkleTree.computeProof(nonNullLeafList, index)

                // Create result map
                val result = Arguments.createMap()
                result.putString("root", MerkleTree.bigIntegerToHex(proof.root))
                result.putInt("index", proof.leafIndex)

                // Convert siblings to hex strings
                val pathArray = Arguments.createArray()
                proof.siblings.forEach { sibling ->
                    pathArray.pushString(MerkleTree.bigIntegerToHex(sibling))
                }
                result.putArray("path", pathArray)

                promise.resolve(result)
            } catch (e: Exception) {
                promise.reject("MERKLE_PROOF_ERROR", "Error computing Merkle proof: ${e.message}")
            }
        }
    }
    
    override fun onCatalystInstanceDestroy() {
        super.onCatalystInstanceDestroy()
        computationScope.cancel()
        synchronized(treeCache) {
            treeCache.clear()
        }
    }
}